File: backend/internal/schemas/book.go
package schemas

type PriceLevelMessage struct {
	PriceTicks int64 `json:"priceTicks"`
	SizeTicks  int64 `json:"sizeTicks"`
}

type BookDeltaMessage struct {
	Type       string `json:"type"`
	Symbol     string `json:"symbol"`
	Seq        uint64 `json:"seq"`
	Side       string `json:"side"` // "bid" | "ask"
	PriceTicks int64  `json:"priceTicks"`
	SizeTicks  int64  `json:"sizeTicks"`
}

func NewBookDeltaMessage(symbol string, seq uint64, side string, priceTicks, sizeTicks int64) BookDeltaMessage {
	return BookDeltaMessage{
		Type: "book_delta", Symbol: symbol, Seq: seq, Side: side,
		PriceTicks: priceTicks, SizeTicks: sizeTicks,
	}
}

type BookSnapshotMessage struct {
	Type   string              `json:"type"`
	Symbol string              `json:"symbol"`
	Seq    uint64              `json:"seq"`
	Bids   []PriceLevelMessage `json:"bids"`
	Asks   []PriceLevelMessage `json:"asks"`
}

func NewBookSnapshotMessage(symbol string, seq uint64, bids, asks []PriceLevelMessage) BookSnapshotMessage {
	return BookSnapshotMessage{Type: "book_snapshot", Symbol: symbol, Seq: seq, Bids: bids, Asks: asks}
}

File: frontend/src/types/book.ts
export interface PriceLevelMessage {
  priceTicks: number;
  sizeTicks: number;
}

export interface BookDeltaMessage {
  type: "book_delta";
  symbol: string;
  seq: number;
  side: "bid" | "ask";
  priceTicks: number;
  sizeTicks: number;
}

export interface BookSnapshotMessage {
  type: "book_snapshot";
  symbol: string;
  seq: number;
  bids: PriceLevelMessage[];
  asks: PriceLevelMessage[];
}

export type BookMessage = BookDeltaMessage | BookSnapshotMessage;
